File: benchmarks/language/variants/r/7_clinvar_filter.R
# Benchmark 7: Real-World ClinVar Variant Analysis
# Filter and classify ClinVar variants from NCBI.

total <- 0L
pathogenic_chroms <- character()
pathogenic_refs <- character()
pathogenic_alts <- character()

con <- file("data_real/clinvar_20k.vcf", "r")
while (length(line <- readLines(con, n = 1)) > 0) {
  if (startsWith(line, "#")) next
  total <- total + 1L
  fields <- strsplit(line, "\t")[[1]]
  chrom <- fields[1]
  ref <- fields[4]
  alt <- fields[5]

  # Parse INFO for CLNSIG
  info_parts <- strsplit(fields[8], ";")[[1]]
  clnsig <- ""
  for (part in info_parts) {
    if (startsWith(part, "CLNSIG=")) {
      clnsig <- sub("CLNSIG=", "", part)
      break
    }
  }

  if (clnsig %in% c("Pathogenic", "Likely_pathogenic", "Pathogenic/Likely_pathogenic")) {
    pathogenic_chroms <- c(pathogenic_chroms, chrom)
    pathogenic_refs <- c(pathogenic_refs, ref)
    pathogenic_alts <- c(pathogenic_alts, alt)
  }
}
close(con)

n_path <- length(pathogenic_chroms)
chr1_path <- sum(pathogenic_chroms %in% c("1", "chr1"))
snps <- sum(nchar(pathogenic_refs) == 1 & nchar(pathogenic_alts) == 1)
indels <- n_path - snps

cat(sprintf("Total variants: %d\n", total))
cat(sprintf("Pathogenic/Likely pathogenic: %d\n", n_path))
cat(sprintf("Pathogenic on chr1: %d\n", chr1_path))
cat(sprintf("Pathogenic SNPs: %d\n", snps))
cat(sprintf("Pathogenic Indels: %d\n", indels))

File: benchmarks/python/7_clinvar_filter.py
"""Benchmark 7: Real-World ClinVar Variant Analysis
Filter and classify ClinVar variants from NCBI."""

total = 0
pathogenic = []

with open("data_real/clinvar_20k.vcf") as f:
    for line in f:
        if line.startswith("#"):
            continue
        total += 1
        fields = line.strip().split("\t")
        chrom = fields[0]
        ref = fields[3]
        alt = fields[4]

        # Parse INFO for CLNSIG
        info = {}
        for item in fields[7].split(";"):
            if "=" in item:
                k, v = item.split("=", 1)
                info[k] = v

        clnsig = info.get("CLNSIG", "")
        if clnsig in ("Pathogenic", "Likely_pathogenic", "Pathogenic/Likely_pathogenic"):
            pathogenic.append((chrom, ref, alt))

chr1_path = [v for v in pathogenic if v[0] in ("1", "chr1")]
snps = sum(1 for _, ref, alt in pathogenic if len(ref) == 1 and len(alt) == 1)
indels = len(pathogenic) - snps

print(f"Total variants: {total}")
print(f"Pathogenic/Likely pathogenic: {len(pathogenic)}")
print(f"Pathogenic on chr1: {len(chr1_path)}")
print(f"Pathogenic SNPs: {snps}")
print(f"Pathogenic Indels: {indels}")
